### PR TITLE
Rollout renku version

### PR DESCRIPTION
**Pull request by pameladelgado**
This PR introduces a new renku-action that creates a PR on terraform-repo with the changes needed to rollout a new renku version.
To be called by the renku repository when tagged for a release (see https://github.com/SwissDataScienceCenter/renku/pull/2644).

To do:

- [x] automatically populate the clusters list
- [x] make one PR per cluster
- [x] clusters to be excluded should be in a list

### CODE REVIEW

**Review by rokroskar — CHANGES_REQUESTED** [review]
Thanks @pameladelgado, this should make rollouts much easier! Please make the list of excluded cluster names configurable. 

A generic question: should this create a single PR/branch or one per cluster? Since it's automated, maybe it's safer to do the latter?

**Comment by pameladelgado** [comment]
> should this create a single PR/branch or one per cluster? Since it's automated, maybe it's safer to do the latter?

This action and its resulting PR was meant for quick releases, i.e. the ones that any member of the team can make and that do not require careful rollout or downtime.

From past experience I think most of the releases are going to rollout to all of the clusters at once. And if we have a 'breaking changes' release, it sufisses to close the automatically generated PR instead of merging it.

**Comment by rokroskar** [comment]
I think most releases require _some_ downtime because the core service cannot be upgraded otherwise. So they will usually probably need some maintenance announcement. If the number of clusters grows, this becomes increasingly difficult to coordinate, so having separate PRs would make sense. But I don't know what the rest of @SwissDataScienceCenter/yat think.

File: rollout-renku-version/rollout-renku-deployment.sh [hunk 1]
```diff
@@ -27,15 +30,14 @@ git config --global user.name "$GIT_USER"
...
-PRODUCTION_DIR="gitops/production"
 
 # use current date
 date=$(date +"%B %dth %Y")
 clusters=$(ls -d ${PRODUCTION_DIR}/*)
 
 for cluster in $clusters
 do
-  if [[ ! $cluster =~ "rancher" && ! $cluster =~ "renku-dev" ]];
+  if [[ ! $cluster =~ $RANCHER_DIR && ! $cluster =~ $DEV_DIR ]];
```
Comment (rokroskar) [inline rollout-renku-version/rollout-renku-deployment.sh:40]: I think we should assume that other clusters could be exempted from automatic updates so I would make it a list and not specific to these two.

**Review by rokroskar — CHANGES_REQUESTED** [review]

**Comment by aledegano** [comment]
> I think most releases require _some_ downtime because the core service cannot be upgraded otherwise. So they will usually probably need some maintenance announcement. If the number of clusters grows, this becomes increasingly difficult to coordinate, so having separate PRs would make sense. But I don't know what the rest of @SwissDataScienceCenter/yat think.

I tend to agree with @rokroskar: I think we are not yet at the point were we are confident rolling out Renku on multiple clusters at once as it requires too much inside knowledge about which services changes and whatnot.

In my opinion before we can really accelerate the Renku deployments we need to go over the chart and ensure that all services implements a blue/green (or similar) deployment strategy.

**Comment by seanrmurphy** [comment]
Yep - I think it's too risky rolling them all out at once.

More generally, I think we do have a kind fundamental inconsistency here: we have lots of tools/infra to support a high cadence of rollouts but if each rollout requires downtime, then this ultimately leads to (in aggregate) quite a lot of downtime. As @aledegano notes, some kind of blue/green approach is really necessary to support this.

**Comment by rokroskar** [comment]
The renku-python team is working on a design for the core service which will make zero-downtime upgrades possible. Right now it's just because of the PV that cannot be mounted to two pods at once that you have this limitation.

**Comment by rokroskar** [comment]
So @pameladelgado it sounds like the consensus is to have 1 PR per deployment. So lets make that change and then this is probably gtg. I've updated the to-do list at the top of the issue.

**Comment by rokroskar** [comment]
_every_ roll out with a new renku-python release requires a downtime because of the core service. And since renku-python is released at the same cadence as renku, that means that every renku release requires a downtime of at least a few minutes. Not a complete downtime, but certain functionality, e.g. making new projects, is not available during the upgrade.

**Comment by pameladelgado** [comment]
I addressed what was left to do, you can see the results in the tf repo:

image

**Pull request merged**